proposals/lip-0053.md

```diff
@@ -252,178 +479,192 @@ crossChainUpdateTransactionParams = {
-  * `newCertificateThreshold` is `params.newCertificateThreshold` if it is non-zero, `partnerValidators.certificateThreshold` otherwise.
-
-#### InboxUpdate Validity
-
-If `params` contains a non-empty `certificate` and an `inboxUpdate`, the validity of the latter is checked by:
-
-* Let `newInboxRoot`, `newInboxAppendPath` and `newInboxSize` be the resulting new root, append path and size of the Merkle tree that would be obtained if all messages from `crossChainMessages` would be appended to the `partnerChannel` inbox. Note that `partnerChannel.inbox` is not updated here.
-* If `inboxUpdate` contains a non-empty `messageWitness`, then update `newInboxRoot` to the output of `calculateRootFromRightWitness(newInboxSize, newInboxAppendPath, inboxUpdate.messageWitness.siblingHashes)` as specified in [LIP 0031](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md).
-* Then validate the newly updated root against the certificate state root using the provided `outboxRootWitness`. Using notation from [LIP 0039][lip-0039], this is done via the function `verify(queryKeys, proof, certificate.stateRoot)` with:
-
-```python
-queryKeys = [outboxKey],
-
-proof = {
-    siblingHashes: outboxRootWitness.siblingHashes,
-    queries: [{
-        key: outboxKey,
-        value: newInboxRoot,
-        bitmap: outboxRootWitness.bitmap
-    }],
-}
-
-outboxKey = MODULE_ID_INTEROPERABILITY
-            || STORE_PREFIX_OUTBOX
-            || SHA-256(uint32be(partnerChainID)).
+def execute(trs: Transaction) -> None:
+    # Verify certificate signature. We do it here because if it fails, the transaction fails rather than being invalid.
+    verifyCertificateSignature(trs)
+    # Process cross-chain messages in inbox update.
+    crossChainMessages = trs.params.inboxUpdate.crossChainMessages
+
+    # First process basic checks for all CCMs.
+    for ccmBytes in crossChainMessages:
+        # Verify general format. Past this point, we can access ccm root properties.
+        try:
+            validateFormat(ccm)
+            # The CCM must come from the sending chain.
+            if trs.params.sendingChainID != ccm.sendingChainID:
+                raise Exception("CCM is not from the sending chain.")
+        except:
+            terminateChain(trs.params.sendingChainID)
+            # In this case, we do not even update the chain account with the new certificate.
+            return
+
+    for ccmBytes in crossChainMessages:
+        try:
+            ccm = decode(crossChainMessageSchema, ccmBytes)
+            # Modules can verify the CCM. In particular, this calls the interoperability verifyCrossChainMessage defined above.
+            # The Token module verifies the escrowed balance in the CCM sending chain.
+            for each module mdl for which verifyCrossChainMessage exists:
+                mdl.verifyCrossChainMessage(trs, ccm)
```

Comment:
This part maybe in common for both `apply` and `forward`, but I wonder if it would not be nicer if all this functionality is defined within `process`/`forward` in the CCM LIP. The 5 processing stages are nicely defined in the rationale above, it then seems inconsistent to split it up as defined here.

Reply:
After reading the CCM LIP, I wonder why the `verifyCrossChainMessage` hook is needed if we already have the `beforeCrossChainCommandExecution` hook directly afterwards. Both hooks are called for all modules, in the first no state changes are allowed in the second one state changes are reverted in case of an exception.

Reply:
Given the last sentence, do you still want to refactor the verification?

Reply:
Conceptually, I would find it cleaner to move the `verifyCrossChainMessage` to the `forward` and `apply` logic so that here we would just have:
```
   for ccmBytes in crossChainMessages:
        appendToInboxTree(trs.params.sendingChainID, ccmBytes)
        ccm = decode(crossChainMessageSchema, ccmBytes)
        # If the receiving chain is the mainchain, apply the CCM.
        if ccm.receivingChainID == CHAIN_ID_MAINCHAIN:
            apply(trs, ccm)
        # If the receiving chain is not the mainchain, forward the CCM.
        elif ccm.receivingChainID != CHAIN_ID_MAINCHAIN:
            forward(trs, ccm)
```

This means from the perspective of CCU execution processing a CCM is a black box that is wrapped in the `apply` and `forward` functions as much as possible. Then for all CCMs the 5 stages mentioned above also happen all in one function.

Reply:
sounds good